File: LeetCode/cpp/122.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        bool bought = false;
        auto size = prices.size();
        int profit = 0, price = 0;
        int i = 0;
        while (i < size) {
            if (!bought) {
                if (prices[i] < prices[i + 1] && i + 1 < size) {
                    price = prices[i];
                    bought = true;
                }
            } else {
                if (prices[i] > prices[i + 1] && i + 1 < size) {
                    profit += (prices[i] - price);
                    price = 0;
                    bought = false;
                }
            }
            ++i;
        }
        if (bought)
            profit += prices[i - 1] - price;
        return profit;
    }
};

File: LeetCode/python3/122.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        size = len(prices)
        bought = False
        profit = 0
        price = 0
        for i in range(0, size - 1):
            if not bought:
                if prices[i] < prices[i + 1]:
                    bought = True
                    price = prices[i]
            else:
                if prices[i] > prices[i + 1]:
                    bought = False
                    profit += prices[i] - price
                    price = 0
        if bought:
            profit += prices[i + 1] - price
        return profit